streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignorTest.java

```diff
@@ -0,0 +1,347 @@
+        createClient(p3, 1);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02);
+        taskAssignor.assign(1);
+
+        assertThat(allActiveTasks(), equalTo(Arrays.asList(task00, task01, task02)));
+        assertThat(allStandbyTasks(), equalTo(Arrays.asList(task00, task01, task02)));
+    }
+
+
+    @Test
+    public void shouldAssignAtLeastOneTaskToEachClientIfPossible() throws Exception {
+        // add a process with 3 threads
+        createClient(p1, 3);
+        // rest only have a single thread
+        createClient(p2, 1);
+        createClient(p3, 1);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02);
+        taskAssignor.assign(0);
+        assertThat(clients.get(p1).assignedTaskCount(), equalTo(1));
+        assertThat(clients.get(p2).assignedTaskCount(), equalTo(1));
+        assertThat(clients.get(p3).assignedTaskCount(), equalTo(1));
+    }
+
+    @Test
+    public void shouldAssignEachActiveTaskToOneClientWhenMoreClientsThanTasks() throws Exception {
+        createClient(p1, 1);
+        createClient(p2, 1);
+        createClient(p3, 1);
+        createClient(4, 1);
+        createClient(5, 1);
+        createClient(6, 1);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02);
+        taskAssignor.assign(0);
+
+        assertThat(allActiveTasks(), equalTo(Arrays.asList(task00, task01, task02)));
+    }
+
+    @Test
+    public void shouldBalanceActiveAndStandbyTasksAcrossAvailableClients() throws Exception {
+        createClient(p1, 1);
+        createClient(p2, 1);
+        createClient(p3, 1);
+        createClient(4, 1);
+        createClient(5, 1);
+        createClient(6, 1);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00, task01, task02);
+        taskAssignor.assign(1);
+
+        for (final ClientState<TaskId> clientState : clients.values()) {
+            assertThat(clientState.assignedTaskCount(), equalTo(1));
+        }
+    }
+
+    @Test
+    public void shouldAssignMoreTasksToClientWithMoreCapacity() throws Exception {
+        createClient(p2, 2);
+        createClient(p1, 1);
+
+        final StickyTaskAssignor<Integer> taskAssignor = createTaskAssignor(task00,
+                                                                            task01,
+                                                                            task02,
+                                                                            new TaskId(1, 0),
+                                                                            new TaskId(1, 1),
+                                                                            new TaskId(1, 2));
+
+        taskAssignor.assign(0);
+        assertTrue("expected client 2 to have more assigned tasks than client 1", clients.get(p2).assignedTaskCount() > clients.get(p1).assignedTaskCount());
```

Comment:
Thinking about this, I guess we could improve `StickyTaskAssinger`. If I am not off, load balancing on stream basis is not optimal -- but I am also not sure if the effort to improve it is worth it...

If we extend this test to assign more tasks, let's say 12, client `p2` will get 7 tasks assigned and `p1` get 5 tasks assigned (while it would be better to assign 8 tasks to `p2` such that all 3 thread get 4 tasks each). The problem is, that the capacity factors are not considered: `p2` should get twice as many tasks assigned as `p1` -- but the algorithm says only "more" -- and this more is determined be the diff of the capacity (ie. in this case `p2` will get at most 2 more tasks assigned than `p1`. 

Or maybe my analysis is wrong (I did not run the code and step through it.)